include/9on12PipelineStateCache.h

```diff
@@ -143,7 +158,15 @@ namespace D3D9on12
         { 
             for (auto &key : m_pPSOKeys)

```

Comment:
I don't know how important this would be, but you could now change this from a set of PSO keys to a vector of `std::weak_ptr<PipelineStateCacheEntry>`. When the PSO gets deleted due to aging out, or one of the subcomponents going away, the `weak_ptr` will get invalidated, so it doesn't get double-deleted when the other subcomponents go away. It means there's only two copies of the key being stored (key in the `unordered_map`, and in the `PipelineStateCacheEntry` struct instead of now 4 copies.

The only downside is that removing entries from the list becomes hard, and so it probably shouldn't be done (often). But a `weak_ptr` is only one pointer, so having stale entries laying around doesn't seem too bad to me. In the common case it'd be less memory due to `vector` not needing `node` headers or dedicated heap allocations for tree nodes.

This is how it works in 11on12 (https://github.com/microsoft/D3D11On12/blob/master/include/PipelineStateCache.hpp#L126). In that codebase, the `vector` is trimmed when new PSOs are added, and additionally the `PipelineStateCacheEntry` stores a *pointer* to the key - specifically, the address of the key stored in the `unordered_map` entry, since that container is guaranteed to be memory stable.

Reply:
I had noticed Chris' original implementation used weak pointers that required cleaning up.  I went with a solution that avoided the use of weak pointers.  I say lets leave the solution in place for now.  If at some point we need to use weak_ptrs, the change can be made.